File: src/java/103-binary-tree-zigzag-level-order-traversal/ZigzagLevelOrder.java
// link		: https://leetcode-cn.com/problems/binary-tree-zigzag-level-order-traversal/
// Author	: Kylin
// Date		: 2021-08-28

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList();
        Queue<TreeNode> queue = new ArrayDeque();

        if (root != null) {
            queue.add(root);
        }

        // 控制入队方向
        boolean left2Right = true;

        while (!queue.isEmpty()) {
            List<Integer> level = new ArrayList();
            int count = queue.size();

            for (int i = 0; i < count; i++) {
                TreeNode node = queue.poll();

				if (left2Right) {
                    level.add(node.val);
                } else {
                    level.add(0, node.val);
                }

                if (node.left != null) {
                    queue.add(node.left);
                }

                if (node.right != null) {
                    queue.add(node.right);
                }
            }
            left2Right = !left2Right;
            result.add(level);
        }

        return result;
    }
}

File: x-go-lc/103-binary-tree-zigzag-level-order-traversal/zigzag_level_order.go
// link		: https://leetcode.cn/problems/binary-tree-level-order-traversal
// Author	: Kylin
// Date		: 2022-12-24

package leetcode

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func zigzagLevelOrder(root *TreeNode) (arrs [][]int) {
	if root == nil {
		return
	}

	queue := []*TreeNode{root}
	for level := 0; len(queue) > 0; level++ {
		var items []int
		for i, length := 0, len(queue); i < length; i++ {
			curNode := queue[0]
			queue = queue[1:]

			items = append(items, curNode.Val)
			if curNode.Left != nil {
				queue = append(queue, curNode.Left)
			}

			if curNode.Right != nil {
				queue = append(queue, curNode.Right)
			}
		}

		if level%2 == 1 {
			items = reverse(items)
		}
		arrs = append(arrs, items)
	}

	return
}

func reverse(arr []int) []int {
	for left, right := 0, len(arr)-1; left < right; {
		arr[left], arr[right] = arr[right], arr[left]
		left++
		right--
	}
	return arr
}
